### PR TITLE
[receiver/prometheus] set pdata stale marker

### LINKED ISSUE #6376: [Prometheus Receiver] Histogram and Summary metric count value outputs minInt64 for non-numerical input values
**Describe the bug**

We want to ensure OpenMetrics / Prometheus compatibility in the OpenTelemetry Collector. We have been building compatibility tests to verify the OpenMetrics spec is fully supported on both the OpenTelemetry Collector Prometheus receiver and PRW exporter as well as in Prometheus itself.

Prometheus Receiver should assign a metric  `staleNaN` value, if the metric is missing in the current scrape but was present in the previous scrape. However, currently metric builder do not assign `staleNaN` values to the histogram and summary values, that are passed by the Prometheus scrape loop for failed scrapes. 
Histogram and Summary `count` values are int64, and the casting of float64 type num-numerical values(staleNaN, normalNaN, and +-Inf) assign minInt64(-9223372036854775808)  number to the `count`. 

**Steps to reproduce**

* Run func [TestEndToEnd(t *testing.T) ](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1e9005e6fa42a5514bac6f4f4c725f10d64c16f8/receiver/prometheusreceiver/metrics_receiver_test.go#L1295)
* Currently, the validate loop is skipped for the tests, re-enable the validate loop by removing/commenting following lines from func testEndToEnd(...) (lines  [1442-1445](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1e9005e6fa42a5514bac6f4f4c725f10d64c16f8/receiver/prometheusreceiver/metrics_receiver_test.go#L1440))

```
if true {
   t.Log(`Skipping the "up" metric checks as they seem to be spuriously failing after staleness marker insertions`)
   return}

```
*  Note: the test fails in getValidScrapes due to staleness, inspect the [metrics](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1e9005e6fa42a5514bac6f4f4c725f10d64c16f8/receiver/prometheusreceiver/metrics_receiver_test.go#L1423) and find the first failed scrape. The Histogram and Summary count values in the failed scrapes are minInt64(-9223372036854775808) instead of non-numerical values (staleNaN, normalNaN, and +-Inf) 

**What did you see instead?**

Scraping endpoints that contains histogram/summary metric and a failed scrape in between, produces the following graph in Prometheus Web UI. The histogram/summary count value is plotted as the peak in the below graph:
Screen Shot 2021-11-01 at 9 40 08 PM

However, if same data is passed directly to the Prometheus Server. Prometheus WebUI produces following graph:
Screen Shot 2021-11-03 at 7 29 30 AM

**Possible Solution**

Since `count` value (int64) for histogram/summary can not be assigned float64 values, one possible solution is to use directly use OTLP format in OTLP Prometheus receiver `metricbuilder`, and assign datapoint flags ([MetricDataPointFlagNoRecordedValue](https://github.com/open-telemetry/opentelemetry-collector/blob/0b4e5ec7924c56342c0ab453ca3dc99bbe3642cd/model/pdata/metrics.go#L285)) to the metric as staleness marker. See linked issue: #6400

**What version did you use?**

Collector-Contrib: v- 0.37.1

**Additional context**
Related to https://github.com/open-telemetry/wg-prometheus/issues/57
Linked Issue: #6400 #6000 #6087

cc @alolita @Aneurysm9

### LINKED ISSUE #6935: [prometheus exporter] Metric value overflow on metrics_expiration
**Describe the bug**
When a metric expires, the value for the metric appears to overflow the max int64 value: `9223372036854775808`.

**Steps to reproduce**
1. Start otel collector with provided config below.
2. Start HotROD example: `docker run   --rm --network="host" --env JAEGER_AGENT_HOST=localhost  --env JAEGER_AGENT_PORT=6835   -p8080-8083:8080-8083   jaegertracing/example-hotrod:latest   all`
    1. This will emit spans to otel collector on port `6835`.
3. Click on a button on HotROD example to send a trace.
4. Check that the `latency_bucket` metrics are correct for the first minute of data, and make a note of which bucket has `count > 0`. For example, where the metric has label `le = "250"`.
    1. Example query: `latency_bucket{service_name = "driver", le="250"}`

**What did you expect to see?**

After a minute (on metric expiry), the metric with `le = "250"` (example query `latency_bucket{service_name = "driver", le="250"}`) should no longer be query-able (a null value).

**What did you see instead?**

After a minute (on metric expiry), the metric with `le = "250"` (example query `latency_bucket{service_name = "driver", le="250"}`) will jump to `9223372036854775808`.

**What version did you use?**
Version: `main` branch

**What config did you use?**
Config: 
```
receivers:
  jaeger:
    protocols:
      thrift_binary:
        endpoint: "localhost:6836"
      thrift_compact:
        endpoint: "localhost:6835"
  otlp/spanmetrics:
    protocols:
      grpc:
        endpoint: :12345
  otlp:
    protocols:
      grpc:
        endpoint: :4317
  prometheus:
    config:
      global:
        external_labels:
          p8s_logzio_name: spm-demo-otel
      scrape_configs: 
      - job_name: 'atm'
        scrape_interval: 15s
        static_configs:
        - targets: [ "0.0.0.0:8889" ]
exporters:
  jaeger:
    endpoint: "localhost:14250"
    tls:
      insecure: true
  prometheus:
    endpoint: "localhost:8889"
    metric_expiration: 1m
  logging:
    loglevel: debug
processors:
  batch:
  spanmetrics:
    metrics_exporter: prometheus
    latency_histogram_buckets: [2ms, 6ms, 10ms, 100ms, 250ms, 500ms, 1000ms, 10000ms, 100000ms, 1000000ms]
    # Additional list of dimensions on top of:
    # - service.name
    # - operation
    # - span.kind
    # - status.code
    dimensions:
      # If the span is missing http.method, the processor will insert
      # the http.method dimension with value 'GET'.
      # For example, in the following scenario, http.method is not present in a span and so will be added as a dimension to the metric with value "GET":
      # - promexample_calls{http_method="GET",operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET"} 1
      - name: http.method
        default: GET
      # If a default is not provided, the http.status_code dimension will be omitted
      # if the span does not contain http.status_code.
      # For example, consider a scenario with two spans, one span having http.status_code=200 and another missing http.status_code. Two metrics would result with this configuration, one with the http_status_code omitted and the other included:
      # - promexample_calls{http_status_code="200",operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET"} 1
      # - promexample_calls{operation="/Address",service_name="shippingservice",span_kind="SPAN_KIND_SERVER",status_code="STATUS_CODE_UNSET"} 1
      - name: http.status_code    
extensions:
  pprof:
    endpoint: :1777
  zpages:
    endpoint: :55679
  health_check:
service:
  extensions: [health_check, pprof, zpages]
  pipelines:
    traces:
      receivers: [jaeger]
      processors: [spanmetrics,batch]
      exporters: [jaeger, logging]
    metrics/spanmetrics:
      # This receiver is just a dummy and never used.
      # Added to pass validation requiring at least one receiver in a pipeline.
      receivers: [otlp/spanmetrics]
      exporters: [prometheus]
    metrics:
      receivers: [otlp,prometheus]
      exporters: [logging]      
  telemetry:
    logs:
      level: "debug"
```

**Environment**
OS: "Ubuntu 20.04"
Compiler(if manually compiled): "go 1.17.5")

**Additional context**

- Is there something the `spanmetrics` processor should be doing to avoid this problem?
- Is it perhaps a config or a promeetheus query issue?
- Is this expected behaviour and is the recommendation to simply increase the `metric_expiration` duration?

The following screenshots illustrate the problem for the default `5m` `metric_expiration` configuration:

**Before Expiry**
before-expiry

**After Expiry**
after-expiry

The logs also reflect the above screenshots, showing the correct metrics initially, then suddenly jumping to a very large count:
- Initially, the `le = "250"` bucket has a count of `3`.
- 5 mins later without sending any traces, a number of logs indicate the expiry of the metrics.
- At this time, `le = "250"` bucket jumps to a very large number, which appears to be 1 + MaxInt64.
- Other buckets remain at 0.

### PR DESCRIPTION
**Pull request by Aneurysm9**
**Description:** Causes the prometheus receiver to set the OTLP "no data present" flag for datapoints representing timeseries where the Prometheus scraper reported that the timeseries is stale.

Removes the old OpenCensus-based metrics adjuster and causes the OpenCensus collection path to use the new pdata-based metrics adjuster as there are subtleties to handling adjustment of stale timeseries that are easier to account for with the pdata-based adjuster.

Fixes #6935.  Fixes #6376.

### CODE REVIEW

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 1]
```diff
@@ -218,19 +218,14 @@ func getValidScrapes(t *testing.T, rms []*pdata.ResourceMetrics) []*pdata.Resour
...
-func isFirstFailedScrape(t *testing.T, metrics []*pdata.Metric) bool {
+func isFirstFailedScrape(metrics []*pdata.Metric) bool {
 	for _, m := range metrics {
 		if m.Name() == "up" {
 			if m.Gauge().DataPoints().At(0).DoubleVal() == 1 { // assumed up will not have multiple datapoints
 				return false
 			}
 		}
 	}
-	// TODO: Issue #6376. Remove this skip once OTLP format is directly used in Prometheus Receiver.
```
Comment (codeboten) [inline receiver/prometheusreceiver/metrics_receiver_helper_test.go:229]: 🎉

File: receiver/prometheusreceiver/internal/transaction.go [hunk 2]
```diff
@@ -193,16 +195,23 @@ func (tr *transaction) Commit() error {
...
+	var md pdata.Metrics
 	if len(metrics) > 0 {
-		md := opencensus.OCToMetrics(tr.node, tr.resource, metrics)
+		md = opencensus.OCToMetrics(tr.node, tr.resource, metrics)
+		fixStaleMetrics(&md)
 		numPoints = md.DataPointCount()
+	}
+
+	if !tr.useStartTimeMetric {
+		// AdjustMetricSlice - jobsMap has to be non-nil in this case.
```
Comment (codeboten) [inline receiver/prometheusreceiver/internal/transaction.go:209]: The comment here refers to the renamed method AdjustMetricsSlice, but the code below still uses AdjustMetrics. Should one of these be updated?

File: receiver/prometheusreceiver/internal/transaction.go [hunk 3]
```diff
@@ -263,3 +272,72 @@ func createNodeAndResource(job, instance, scheme string) (*commonpb.Node, *resou
 	}
 	return node, resource
 }
+
+func fixStaleMetrics(md *pdata.Metrics) {
```
Comment (codeboten) [inline receiver/prometheusreceiver/internal/transaction.go:276]: Are the current tests sufficient to cover the cases from this function?

File: receiver/prometheusreceiver/internal/transaction.go (hunk 2 above)
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/transaction.go:209]: Good catch, is a remnant of automated refactor that renamed the prior `AdjustMetrics` implementation to `AdjustMetricSlice`.  I have removed these comments as this one adds no value and the one below it is outdated and incorrect.

File: receiver/prometheusreceiver/internal/transaction.go (hunk 3 above)
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/transaction.go:274]: I believe so.  There are end-to-end test for all aggregation types that include responses that should induce staleness markers.  The tests that are enabled in this PR ensure that those staleness markers are generated and that start time adjustment is correctly handled when they are present.

**Review by codeboten — APPROVED** [review]
Pinging @dashpole as the other code owner

**Review by dashpole — APPROVED** [review]

**Pull request merged**